### PR TITLE
Disable payments for GKE

### PR DESCRIPTION
**Pull request by rmehner**
I think I've adapted all the relevant jobs.

Couple of notes:

1. I've opted to not just check in `lib/payments`, but rather have it explicitly in all the jobs. It's more code, but I like that it's explicit there.

2. While I refactored a little bit, I didn't do too much. If we ever have a bit more time, we should refactor the tests more thoroughly.

### CODE REVIEW

**Comment by Realtin** [comment]
Cool!  Can you tell me more about the use of requireFresh in some tests? why is it needed in just some of them and not in others?

**Comment by rmehner** [comment]
> Cool! Can you tell me more about the use of requireFresh in some tests? why is it needed in just some of them and not in others?

Sure :) It is relevant to those tests that fiddle with the environment. We're using [envalid][1] to have our `process.env` in a sanitized way. The problem is that if we require a module that requires our [./lib/env][2], any subsequent change to `process.env` won't be seen, as the code in the mentioned [./lib/env][2] only runs once. Even if we re-require the module, it won't work, since node has a module cache that will be hit during the second require and the code won't be execute a second time.

One way to mitigate this would be to use `process.env` directly, but then we'd lose all the niceness that we get through [envalid][1].

The other way is to clear the cache for the file before you require it again.

If you look at [module-cache-helpers][3] you see those two functions: `clearCache` and `requireFresh`. With `clearCache` you can clear the cache for file. In our case we use that to clear the cache for [./lib/env][2]. `requireFresh` clears the cache it is about to require and then requires the file. This allows us to run the tests with a fresh environment whenever we needed.

Hope this clears things up a bit, happy to answer any questions :)

[1]: https://github.com/af/envalid
[2]: https://github.com/greenkeeperio/greenkeeper/blob/master/lib/env.js
[3]: https://github.com/greenkeeperio/greenkeeper/blob/features/gke-2.0/test/helpers/module-cache-helpers.js

File: jobs/create-version-branch.js [hunk 1]
```diff
@@ -86,7 +86,7 @@ module.exports = async function (
     return
   }
 
-  if (repository.private) {
+  if (repository.private && !env.IS_ENTERPRISE) {
```
Comment (janl) [inline jobs/create-version-branch.js]: 🎨 in https://github.com/greenkeeperio/greenkeeper/pull/602/commits/a13f03b905faaa9eb1a6e6112bf1ba3619c0fb8a#diff-137e495c3f0b01ee061147fede39cbacR67 you put the `IS_ENTERPRISE` check to the front. We should agree on a uniform way, which one doesn’t matter.

File: test/jobs/github-event/pull_request/closed.js [hunk 2]
```diff
@@ -60,12 +79,51 @@ test('github-event pull_request closed', async t => {
     t.ok(repository.enabled, 'repository is enabled')
     t.end()
   })
+
+  t.test('initial pr merged on private repo with payment plan', async t => {
```
Comment (janl) [inline test/jobs/github-event/pull_request/closed.js]: ❓ why did you add this test?

File: test/jobs/create-initial-branch.js [hunk 3]
```diff
@@ -162,6 +168,52 @@ test('create-initial-branch', async t => {
     t.end()
   })
 
+  t.test('badge already added for private repo', async t => {
```
Comment (janl) [inline test/jobs/create-initial-branch.js]: ❓ why did you add this test?

File: test/jobs/github-event/pull_request/opened.js [hunk 4]
```diff
@@ -91,17 +92,10 @@ test('github-event pull_request opened', async t => {
 
 tearDown(async () => {
   const { repositories } = await dbs()
-  await repositories.remove(await repositories.get('42:pr:666'))
-  await repositories.remove(await repositories.get('42'))
-  const docIds = [667, 668]
+  await removeIfExists(repositories, '42')
```
Comment (janl) [inline test/jobs/github-event/pull_request/opened.js]: 👍

File: jobs/github-event/pull_request/opened.js [hunk 5]
```diff
@@ -36,7 +37,7 @@ module.exports = async function (data) {
     }
   )
 
-  if (!repoDoc.private) {
+  if (!repoDoc.private || env.IS_ENTERPRISE) {
```
Comment (janl) [inline jobs/github-event/pull_request/opened.js]: 🎨 see earlier, `IS_ENTERPRISE` location in `if` clause

File: test/jobs/github-event/pull_request/opened.js [hunk 6]
```diff
@@ -53,11 +68,48 @@ test('github-event pull_request opened', async t => {
     t.end()
   })
 
+  t.test('initial pr on private repo opened', async t => {
```
Comment (janl) [inline test/jobs/github-event/pull_request/opened.js]: ❓ why did you add this test?

File: test/jobs/github-event/push.js [hunk 7]
```diff
@@ -379,16 +392,63 @@ test('github-event push', async t => {
     t.notOk(repo.enabled)
     t.end()
   })
+
+  t.test('package.json deleted on private repo', async t => {
```
Comment (janl) [inline test/jobs/github-event/push.js]: ❓ why did you add this test?

File: test/jobs/reset.js [hunk 8]
```diff
@@ -66,13 +66,8 @@ test('reset repo', async t => {
...
   t.afterEach(async () => {
     nock.cleanAll()
-    await removeIfExists(repositories, '42')
-    await installations.remove(await installations.get('123'))
-    await removeIfExists(repositories, '42:pr:123')
-    await removeIfExists(repositories, '42:branch:deadbeef')
-    await removeIfExists(repositories, '42:branch:deadbeef0')
-    await removeIfExists(repositories, '42:issue:67')
-    await removeIfExists(repositories, '42:issue:65')
+    await removeIfExists(repositories, '42', '42:pr:123', '42:branch:deadbeef', '42:branch:deadbeef0', '42:issue:67', '42:issue:65')
```
Comment (janl) [inline test/jobs/reset.js]: :art: line gets a little long, no biggie overall, but we might wanna think about a default format with more line breaks, or passing an array that we construct over multiple lines.

File: test/jobs/github-event/pull_request/closed.js (hunk 2 above)
Comment (rmehner) [reply test/jobs/github-event/pull_request/closed.js]: It wasn't tested before and since the special enterprise handling affects private repos, I wanted to have this as a test first, so I can check that the behaviour for private repos on enterprise is different.

File: test/jobs/create-initial-branch.js (hunk 3 above)
Comment (rmehner) [reply test/jobs/create-initial-branch.js]: Same as https://github.com/greenkeeperio/greenkeeper/pull/602#discussion_r164702375

File: test/jobs/github-event/pull_request/opened.js (hunk 6 above)
Comment (rmehner) [reply test/jobs/github-event/pull_request/opened.js]: Same as https://github.com/greenkeeperio/greenkeeper/pull/602#discussion_r164702375

File: test/jobs/github-event/push.js (hunk 7 above)
Comment (rmehner) [reply test/jobs/github-event/push.js]: Same as https://github.com/greenkeeperio/greenkeeper/pull/602#discussion_r164702375 :D

File: test/jobs/github-event/pull_request/closed.js (hunk 2 above)
Comment (janl) [reply test/jobs/github-event/pull_request/closed.js]: Thanks!

File: jobs/create-version-branch.js (hunk 1 above)
Comment (rmehner) [reply jobs/create-version-branch.js]: Will do. I'd opt for `private` first, as this allows for some early exits and most of the special behaviour only affects private repos anyway.

Comment (janl) [reply jobs/create-version-branch.js]: I like that. In addition, that’s the main code logic, and Enterprise is a secondary consideration, so it works well at the end.

File: test/jobs/reset.js (hunk 8 above)
Comment (rmehner) [reply test/jobs/reset.js]: In bffa40b I refactored `removeIfExists` to also take an array, in 899a57e I'm using it to make the line more readable :)

File: jobs/create-version-branch.js (hunk 1 above)
Comment (rmehner) [reply jobs/create-version-branch.js]: b771be9 done

**Comment by rmehner** [comment]
Also rebased & `--force-with-lease`. LGTM now.
